RecoLocalCalo/HGCalRecAlgos/src/HGCalImagingAlgo.cc

```diff
@@ -167,20 +167,39 @@ math::XYZPoint HGCalImagingAlgo::calculatePosition(std::vector<KDNode> &v){
   float y = 0.;
   float z = 0.;
   unsigned int v_size = v.size();
+  unsigned int maxEnergyIndex = 0;
+  float maxEnergyValue = 0;
+  bool haloOnlyCluster = true;
 
+  // loop over hits in cluster candidate building up weight for
+  // energy-weighted position calculation and determining the maximum
+  // energy hit in case this is a halo-only cluster
   for (unsigned int i = 0; i < v_size; i++){
     if(!v[i].data.isHalo){
+      haloOnlyCluster = false;
       total_weight += v[i].data.weight;
       x += v[i].data.x*v[i].data.weight;
       y += v[i].data.y*v[i].data.weight;
       z += v[i].data.z*v[i].data.weight;
     }
+    else {
+      if (v[i].data.weight > maxEnergyValue) {
+        maxEnergyValue = v[i].data.weight;
+        maxEnergyIndex = i;
+      }
+    }
   }
 
-  if (total_weight != 0) {
-  return math::XYZPoint( x/total_weight,
+  if (!haloOnlyCluster) {
+    if (total_weight != 0) {
+      return math::XYZPoint( x/total_weight,
 			 y/total_weight,
 			 z/total_weight );
+    }
+  }
+  else {
+    // return position of hit with maximum energy
+    return math::XYZPoint(v[maxEnergyIndex].data.x, v[maxEnergyIndex].data.y, v[maxEnergyIndex].data.z);
```

Comment:
On the off-chance that `v.empty()`, we'll still have `maxEnergyIndex=0`, but `v[maxEnergyIndex]` won't exist. There should be a protection for this.